Re-sync color conversion constants with latest in spec

weinig opened the pull request:
#### 1d979e658a4948cb03f593c9091b70b275eff5c3
<pre>
Re-sync color conversion constants with latest in spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=272952">https://bugs.webkit.org/show_bug.cgi?id=272952</a>

Reviewed by Darin Adler.

Update the following constants:
- xyzToLinear/linearToXYZ 3x3 matrices for the RGB color types.
- chromatic adaptation 3x3 matrices
- D50 and D65 white point vector
- OKLab XYZ &lt;-&gt; LMS matrices

Also updated polar/rectilinear conversions to correctly
handle the achromatic cases, ensuring conversion to lch/oklch
have &quot;powerless&quot; hue when chroma is close to 0, and when
that &quot;powerless&quot; hue is transformed back to lab/oklab, that
it the opponent A/B values are set to 0.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
Update one result to account to more closely match colorjs.io reference results.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/nested-color-mix-with-currentcolor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    - Updated results that are now passing.

* Tools/TestWebKitAPI/Tests/WebCore/ColorTests.cpp:
(TestWebKitAPI::TEST(Color, P3ConversionToSRGB)):
    - Updated result to account for change in OKLab conversion. (OKLab is used for the
      gamut mapping, which is why this only effected this one test).

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertToPolarForm):
    - Updated to account for sufficiently small a/b values that the result would be achromatic.

(WebCore::convertToRectangularForm):
    - Handle missing (NaN) hue, to correctly set a/b to 0.

(WebCore::OKLab&lt;float&gt;&gt;::convert):
    - Update XYZ &lt;-&gt; LMS constants to latest in spec.

* Source/WebCore/platform/graphics/ColorConversion.h:
    - Move and update white point constants to the header to be closer to the chromatic adaptation code.
    - Update chromatic adaptation constants and add a comment explaining what they are all about.

* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::ColorMatrix::ColorMatrix):
    - Allow conversions on construction to allow matrices to be defined with either double or float.

(WebCore::brightnessColorMatrix):
(WebCore::contrastColorMatrix):
(WebCore::invertColorMatrix):
(WebCore::opacityColorMatrix):
    - Fix a typo that was copied in a few places. compontent -&gt; component

* Source/WebCore/platform/graphics/ColorTypes.h:
    - Update LinearRGB &lt;-&gt; XYZ constants.

Canonical link: <a href="https://commits.webkit.org/278112@main">https://commits.webkit.org/278112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52ed0f3bba9dcb4b6f6b0096a0f592cbc790a75b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49532 "37 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52576 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52778 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51836 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34832 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26434 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40448 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26357 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21563 "Found 1 new API test failure: /TestWebCore:Color.P3ConversionToSRGB (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23822 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43856 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7908 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45732 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54350 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24619 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20808 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25887 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46837 "Found 1 new API test failure: /TestWebCore:Color.P3ConversionToSRGB (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10876 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26735 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25609 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -42,8 +42,13 @@ LCHLike convertToPolarForm(const LabLike& color)
+    static constexpr auto epsilon = 0.02f;
+
     float chroma = std::hypot(a, b);
+    float hue = (std::abs(a) < epsilon && std::abs(b) < epsilon) ? std::numeric_limits<float>::quiet_NaN() : rad2deg(atan2(b, a));
```
Does this use the `double` versions of `rad2deg` and `atan2`? If so, is that better than using the `float` versions?

weinig replied:
I'm not sure, but I am going to be doing an experiment of making all the conversion code use double for everything, delaying conversion back to float until all substeps / transforms are completed, so it will be a good opportunity to look into this.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -199,9 +207,8 @@ ExtendedSRGBA<float> ColorConversion<ExtendedSRGBA<float>, HWBA<float>>::convert
-static constexpr float LABk = 24389.0f / 27.0f;
-static constexpr float D50WhiteValues[] = { 0.96422f, 1.0f, 0.82521f };
+static constexpr float LABe = 216.0 / 24389.0;
+static constexpr float LABk = 24389.0 / 27.0;
```
Pretty sure these `static` have no effect and can be omitted.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -281,15 +288,15 @@ XYZA<float, WhitePoint::D65> ColorConversion<XYZA<float, WhitePoint::D65>, OKLab
     // https://bottosson.github.io/posts/oklab/ with XYZ <-> LMS matrices recalculated for consistent reference white in https://github.com/w3c/csswg-drafts/issues/6642#issuecomment-943521484
 
     static constexpr ColorMatrix<3, 3> LinearLMSToXYZD65 {
```
This `static` may have no effect and if so we should omit it.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -281,15 +288,15 @@ XYZA<float, WhitePoint::D65> ColorConversion<XYZA<float, WhitePoint::D65>, OKLab
+        -0.0763729366746601, -0.4214933324022432,  1.5869240198367816,
     };
 
     static constexpr ColorMatrix<3, 3> OKLabToNonLinearLMS {
```
This `static` may have no effect and if so we should omit it.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -317,15 +324,15 @@ OKLab<float> ColorConversion<OKLab<float>, XYZA<float, WhitePoint::D65>>::conver
     // https://bottosson.github.io/posts/oklab/ with XYZ <-> LMS matrices recalculated for consistent reference white in https://github.com/w3c/csswg-drafts/issues/6642#issuecomment-943521484
 
     static constexpr ColorMatrix<3, 3> XYZD65ToLinearLMS {
```
This `static` may have no effect and if so we should omit it.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.cpp:
```diff
@@ -317,15 +324,15 @@ OKLab<float> ColorConversion<OKLab<float>, XYZA<float, WhitePoint::D65>>::conver
+        0.0481771893596242,  0.2642395317527308,  0.6335478284694309,
     };
 
     static constexpr ColorMatrix<3, 3> NonLinearLMSToOKLab {
```
This `static` may have no effect and if so we should omit it.

darinadler commented on Source/WebCore/platform/graphics/ColorConversion.h:
```diff
@@ -61,25 +61,73 @@ template<typename Output, typename Input> inline Output convertColor(const Input
+// MARK: White Point.
+
+inline constexpr float const D50WhitePoint[] = { 0.3457 / 0.3585, 1.0, (1.0 - 0.3457 - 0.3585) / 0.3585 };
+inline constexpr float const D65WhitePoint[] = { 0.3127 / 0.3290, 1.0, (1.0 - 0.3127 - 0.3290) / 0.3290 };
```
What does the `inline` do here?

We don't need the `const` and it should be omitted.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
You'll need to export this fwiw (unless this is syncing back from upstream)

weinig replied:
Yep, will do.

Commit:
Re-sync color conversion constants with latest in spec
https://bugs.webkit.org/show_bug.cgi?id=272952

Reviewed by Darin Adler.

Update the following constants:
- xyzToLinear/linearToXYZ 3x3 matrices for the RGB color types.
- chromatic adaptation 3x3 matrices
- D50 and D65 white point vector
- OKLab XYZ <-> LMS matrices

Also updated polar/rectilinear conversions to correctly
handle the achromatic cases, ensuring conversion to lch/oklch
have "powerless" hue when chroma is close to 0, and when
that "powerless" hue is transformed back to lab/oklab, that
it the opponent A/B values are set to 0.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color.html:
Update one result to account to more closely match colorjs.io reference results.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/nested-color-mix-with-currentcolor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-computed-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-relative-color-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
    - Updated results that are now passing.

* Tools/TestWebKitAPI/Tests/WebCore/ColorTests.cpp:
(TestWebKitAPI::TEST(Color, P3ConversionToSRGB)):
    - Updated result to account for change in OKLab conversion. (OKLab is used for the
      gamut mapping, which is why this only effected this one test).

* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::convertToPolarForm):
    - Updated to account for sufficiently small a/b values that the result would be achromatic.

(WebCore::convertToRectangularForm):
    - Handle missing (NaN) hue, to correctly set a/b to 0.

(WebCore::OKLab<float>>::convert):
    - Update XYZ <-> LMS constants to latest in spec.

* Source/WebCore/platform/graphics/ColorConversion.h:
    - Move and update white point constants to the header to be closer to the chromatic adaptation code.
    - Update chromatic adaptation constants and add a comment explaining what they are all about.

* Source/WebCore/platform/graphics/ColorMatrix.h:
(WebCore::ColorMatrix::ColorMatrix):
    - Allow conversions on construction to allow matrices to be defined with either double or float.

(WebCore::brightnessColorMatrix):
(WebCore::contrastColorMatrix):
(WebCore::invertColorMatrix):
(WebCore::opacityColorMatrix):
    - Fix a typo that was copied in a few places. compontent -> component

* Source/WebCore/platform/graphics/ColorTypes.h:
    - Update LinearRGB <-> XYZ constants.

Canonical link: https://commits.webkit.org/278112@main
